deploy-service/teletraanservice/src/main/java/com/pinterest/teletraan/worker/AgentJanitor.java

```diff
@@ -108,21 +121,40 @@ private void processHighWatermarkHosts() throws Exception {
         long current_time = System.currentTimeMillis();
         long maxThreshold = current_time - Math.min(maxStaleHostThreshold, maxLaunchLatencyThreshold);
         List<HostAgentBean> maxStaleHosts = hostAgentDAO.getStaleHosts(maxThreshold);
-        Set<String> staleHostIds = new HashSet<>();
+        //Set<String> staleHostIds = new HashSet<>();
+        Map<String, String> staleHostIds = new HashMap<>();
         
         for (HostAgentBean hostAgentBean : maxStaleHosts) {
             if (isHostStale(hostAgentBean)) {
-                staleHostIds.add(hostAgentBean.getHost_id());
+                staleHostIds.put(hostAgentBean.getHost_id(), hostAgentBean.getAuto_scaling_group());
             }
         }
-        Collection<String> terminatedHosts = getTerminatedHostsFromSource(staleHostIds);
-        for (String staleId : staleHostIds) {
+        Collection<String> terminatedHosts = getTerminatedHostsFromSource(staleHostIds.keySet());
+        Map<String, Integer> unreachableCountPerASG = new HashMap<>();
+        for (Map.Entry<String, String> entry : staleHostIds.entrySet()) {
+            String staleId = entry.getKey();
             if (terminatedHosts.contains(staleId)) {
                 removeStaleHost(staleId);
             } else {
                 markUnreachableHost(staleId);
+                unreachableCountPerASG.put(entry.getValue(), unreachableCountPerASG.getOrDefault(entry.getValue(), 0) + 1);
             }
         }
+        notifyOnUnreachableHosts(unreachableCountPerASG);
+    }
+
+    private void notifyOnUnreachableHosts(Map<String, Integer> unreachableCountPerASG) throws Exception {
+        for (Map.Entry<String, Integer> entry : unreachableCountPerASG.entrySet()) {
```

Comment:
maybe add a protection on the number of notification can send so not to spam/overload email/slack server?  for example, if ASG map size > 20 (?), then only notifiy teletraan team.

Reply:
After thinking a bit more, I think we probably don't want to slack notify on unreachable instances but rather export this as metric and have user setup alerts. Along with that, we can notify Teletraan on large number of unreachable instances.
I am thinking of closing this PR and instead work on metrics.